Some Fixes from circular17 suggestions to GetPart #297

maxm74 opened the pull request.

maxm74 committed:
Added TCustomUniversalBitmap.GetPart overloaded function and ATile default Param

Added TPhysicalRect type;
Added Rect arrays Types;
TCustomUniversalBitmap: Added ATile default Param to GetPart;
Added GetPart overloaded function with TPhysicalRect and arrays;

maxm74 committed:
Changed TPhysicalRect unit of measurement to TPhysicalUnit; Added TPhysicalRect Width/Height;

Changed TPhysicalRect unit of measurement to TPhysicalUnit;
Added TPhysicalRect Width/Height;

maxm74 committed:
Changed TPhysicalRect.PhysicalUnit to TCSSUnit; Added PhysicalSizeToPixels overloads with RectF and Rect;

Changed TPhysicalRect.PhysicalUnit to TCSSUnit;
Added PhysicalSizeToPixels overloads with RectF and Rect;

maxm74 committed:
Added PhysicalSizeConvert functions

Added PhysicalSizeConvert functions

maxm74 committed:
Fixed PhysicalSizeConvert with ASourceUnit = ATargetUnit return nothing

maxm74 committed:
Fix Resample Larger Assign #303

maxm74 committed:
Fix Resample Larger Assign bgrabitmap #303 part 2

maxm74 committed:
Some Fixes from circular17 suggestions

Removed unnecessary overloaded GetPart methods;
Use PutImage in the GetPart method with ATile False;
Renamed PhysicalSizeToPixels to PhysicalSizeToRoundedPixels when rounded;
Added PreservesMoreData param to PhysicalSizeToRoundedPixels.

circular17 commented on bgrabitmap/bgraunits.pas:
```diff
@@ -201,6 +231,48 @@ procedure PhysicalSizeToPixels(var SizeX, SizeY: Single;
+begin
+  resRect:= PhysicalSizeToPixels(APhysicalRect, ABitmap);
+
+  if PreservesMoreData
```
It is a good idea to have a parameter to determine the rounding method.

To take into account negative values, I suggest to use `Floor` instead of `Trunc` for Left/Top.

And to include all pixels touched by the rectangle, to use `Ceil` for Right/Bottom as well when `PreservesMoreData` is True.

Finally, one need to consider if the rectangle is flipped. When Left is greater than Right (respectively when Top is greater than Bottom), the reverse rounding needs to be applied.

So that would be something like :

```pascal
function PhysicalSizeToRoundedPixels(const APhysicalRect: TPhysicalRect;
                                     ABitmap: TCustomUniversalBitmap;
                                     PreservesMoreData: Boolean=False): TRect;
var
  resRect: TRectF;
begin
  resRect := PhysicalSizeToPixels(APhysicalRect, ABitmap);

  if PreservesMoreData then
  begin
    // Preserve as much data as possible:
    // use Floor for min coordinates, Ceil for max coordinates
    if resRect.Left <= resRect.Right then
    begin
      Result.Left   := Floor(resRect.Left);
      Result.Right  := Ceil(resRect.Right);
    end
    else
    begin
      // flipped horizontally: invert rounding
      Result.Left   := Ceil(resRect.Left);
      Result.Right  := Floor(resRect.Right);
    end;

    if resRect.Top <= resRect.Bottom then
    begin
      Result.Top    := Floor(resRect.Top);
      Result.Bottom := Ceil(resRect.Bottom);
    end
    else
    begin
      // flipped vertically: invert rounding
      Result.Top    := Ceil(resRect.Top);
      Result.Bottom := Floor(resRect.Bottom);
    end;
  end
  else
  begin
    // Default: round half up (closest integer)
    Result.Left   := HalfUp(resRect.Left);
    Result.Top    := HalfUp(resRect.Top);
    Result.Right  := HalfUp(resRect.Right);
    Result.Bottom := HalfUp(resRect.Bottom);
  end;
end;
```

maxm74 replied:
Added to Code

circular17 reviewed (changes requested):
Thank you for taking into account my remarks in the other pull requests.

I have a few more suggestions. Please have a look.

Note: you can make the changes in the same branch, the pull request will be updated automatically.

maxm74 commented:
In this branch (fix-resample-assign) there is only the GetPart code, plus a
small fix that is always related to GetPart.
I wrote the code without thinking directly on the dev branch, then I
created this branch.
Just consider this one.



Il Sab 13 Set 2025, 10:20 circular17 ***@***.***> ha scritto:

> ***@***.**** requested changes on this pull request.
>
> Thank you for taking into account my remarks in the other pull requests.
>
> I have a few more suggestions. Please have a look.
>
> Note: you can make the changes in the same branch, the pull request will
> be updated automatically.
> ------------------------------
>
> In bgrabitmap/unibitmap.inc
> <https://github.com/bgrabitmap/bgrabitmap/pull/310#discussion_r2346032633>
> :
>
> >    result.InvalidateBitmap;
>  end;
>
> +procedure TCustomUniversalBitmap.GetPart(const ARects: TRectArray; var ABitmaps: TCustomUniversalBitmapArray;
> +                                         ACopyProperties: Boolean; ATile: Boolean);
> +var
> +   i, last: Integer;
> +   curBitmap: TCustomUniversalBitmap;
> +
> +begin
> +  for i:=Low(ARects) to High(ARects) do
> +  begin
> +    curBitmap:= GetPart(ARects[i], ACopyProperties, ATile);
> +    if (curBitmap <> nil) then
>
> What I meant is that the result array should have the same size as the
> parameter array. Otherwise, you might get indexing errors or lose track of
> which result corresponds to which parameter. So I suggest to remove the
> "if" check.
> ------------------------------
>
> In bgrabitmap/unibitmap.inc
> <https://github.com/bgrabitmap/bgrabitmap/pull/310#discussion_r2346032832>
> :
>
> > +                                        ACopyProperties: Boolean; ATile: Boolean): TCustomUniversalBitmap;
> +begin
> +  Result:= GetPart(PhysicalSizeToRoundedPixels(ARect, Self), ACopyProperties, ATile);
> +end;
> +
> +procedure TCustomUniversalBitmap.GetPart(const ARects: TPhysicalRectArray; var ABitmaps: TCustomUniversalBitmapArray;
> +                                         ACopyProperties: Boolean; ATile: Boolean);
> +var
> +   i, last: Integer;
> +   curBitmap: TCustomUniversalBitmap;
> +
> +begin
> +  for i:=Low(ARects) to High(ARects) do
> +  begin
> +    curBitmap:= GetPart(ARects[i], ACopyProperties, ATile);
> +    if (curBitmap <> nil) then
>
> Same here
> ------------------------------
>
> In bgrabitmap/bgraunits.pas
> <https://github.com/bgrabitmap/bgrabitmap/pull/310#discussion_r2346057406>
> :
>
> > +    result.Top := PhysicalSizeToPixels(APhysicalRect.Top, res.ResolutionUnit, res.ResolutionY, APhysicalRect.PhysicalUnit);
> +    result.Left := PhysicalSizeToPixels(APhysicalRect.Left, res.ResolutionUnit, res.ResolutionX, APhysicalRect.PhysicalUnit);
> +    result.Bottom := PhysicalSizeToPixels(APhysicalRect.Bottom, res.ResolutionUnit, res.ResolutionY, APhysicalRect.PhysicalUnit);
> +    result.Right := PhysicalSizeToPixels(APhysicalRect.Right, res.ResolutionUnit, res.ResolutionX, APhysicalRect.PhysicalUnit);
> +  end;
> +end;
> +
> +function PhysicalSizeToRoundedPixels(const APhysicalRect: TPhysicalRect; ABitmap: TCustomUniversalBitmap;
> +                                     PreservesMoreData: Boolean=False): TRect;
> +var
> +   resRect: TRectF;
> +
> +begin
> +  resRect:= PhysicalSizeToPixels(APhysicalRect, ABitmap);
> +
> +  if PreservesMoreData
>
> It is a good idea to have a parameter to determine the rounding method.
>
> To take into account negative values, I suggest to use Floor instead of
> Trunc for Left/Top.
>
> And to include all pixels touched by the rectangle, to use Ceil for
> Right/Bottom as well when PreservesMoreData is True.
>
> Finally, one need to consider if the rectangle is flipped. When Left is
> greater than Right (respectively when Top is greater than Bottom), the
> reverse rounding needs to be applied.
>
> So that would be something like :
>
> function PhysicalSizeToRoundedPixels(const APhysicalRect: TPhysicalRect;
>                                      ABitmap: TCustomUniversalBitmap;
>                                      PreservesMoreData: Boolean=False): TRect;var
>   resRect: TRectF;begin
>   resRect := PhysicalSizeToPixels(APhysicalRect, ABitmap);
>
>   if PreservesMoreData then
>   begin
>     // Preserve as much data as possible:
>     // use Floor for min coordinates, Ceil for max coordinates
>     if resRect.Left <= resRect.Right then
>     begin
>       Result.Left   := Floor(resRect.Left);
>       Result.Right  := Ceil(resRect.Right);
>     end
>     else
>     begin
>       // flipped horizontally: invert rounding
>       Result.Left   := Ceil(resRect.Left);
>       Result.Right  := Floor(resRect.Right);
>     end;
>
>     if resRect.Top <= resRect.Bottom then
>     begin
>       Result.Top    := Floor(resRect.Top);
>       Result.Bottom := Ceil(resRect.Bottom);
>     end
>     else
>     begin
>       // flipped vertically: invert rounding
>       Result.Top    := Ceil(resRect.Top);
>       Result.Bottom := Floor(resRect.Bottom);
>     end;
>   end
>   else
>   begin
>     // Default: round half up (closest integer)
>     Result.Left   := HalfUp(resRect.Left);
>     Result.Top    := HalfUp(resRect.Top);
>     Result.Right  := HalfUp(resRect.Right);
>     Result.Bottom := HalfUp(resRect.Bottom);
>   end;end;
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bgrabitmap/bgrabitmap/pull/310#pullrequestreview-3219863237>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABWJVZZREXHQRWAYD7BAFB33SPHVXAVCNFSM6AAAAACGKWEDNCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTEMJZHA3DGMRTG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

maxm74 committed:
Fixed Gaps in GetPart array result

maxm74 commented:
Testing the code, I saw that TCustomUniversalBitmap.GetPart does not normalize the Rectangle, so if I deliberately pass a rectangle with Left>Right, the result is an image with Width=0. 
It would be a good idea to Normalize the rectangle.

maxm74 committed:
Added PreserveMoreData param to GetPart; Use Floor/Ceil in PhysicalSizeToRoundedPixels

maxm74 committed:
Deleted SizeConv consts and use of InchFactor intestead; Added CSSUnit Long Names;

Deleted SizeConv consts and use of InchFactor intestead;
Added CSSUnit Long Names;

maxm74 committed:
TCustomUniversalBitmap.GetPart Added Normalize the Rect

circular17 reviewed (approved):
Looks good 👍 

I'll check the `ClipRect` assignment change.
